db/blip_sync_context.go

```diff
@@ -617,8 +620,26 @@ func (bsc *BlipSyncContext) sendNoRev(sender *blip.Sender, docID, revID string,
 // Pushes a revision body to the client
 func (bsc *BlipSyncContext) sendRevision(sender *blip.Sender, docID, revID string, seq SequenceID, knownRevs map[string]bool, maxHistory int, handleChangesResponseCollection *DatabaseCollectionWithUser, collectionIdx *int) error {
 	rev, err := handleChangesResponseCollection.GetRev(bsc.loggingCtx, docID, revID, true, nil)
+
+	var replacedRevID string
+
+	collectionCtx, collectionErr := bsc.collections.get(collectionIdx)
+	if collectionErr != nil {
+		return collectionErr
+	}
+
+	if collectionCtx.sendReplacementRevs && base.IsDocNotFoundError(err) {
+		base.DebugfCtx(bsc.loggingCtx, base.KeySync, "Unavailable revision for %q %s - finding replacement: %v", base.UD(docID), revID, err)
+
+		// try the active rev instead as a replacement
+		replacementRev, replacementRevErr := handleChangesResponseCollection.GetRev(bsc.loggingCtx, docID, "", true, nil)
+		// set replacement and continue as normal
+		replacedRevID = revID
+		revID = replacementRev.RevID
+		rev, err = replacementRev, replacementRevErr
+	}
 	if base.IsDocNotFoundError(err) {
-		return bsc.sendNoRev(sender, docID, revID, collectionIdx, seq, err)
+		return bsc.sendNoRev(sender, docID, revID, replacedRevID, collectionIdx, seq, err)
```

Comment:
I'm not sure under what scenarios where we want to send replacedRevID on a noRev (are there any?), but IsDocNotFound seems like a clear case where we don't have a replacedRevID to send, but we could fall into this after also matching the previous sendReplacementRevs check (if line 635 returns not found).

Reply:
Agreed. This is very poorly structured - I tried to reuse existing flow as much as possible but I think it actually complicates it more. Will redo replacement rev sending to be more explicit.

Reply:
Updated to be explicit and align with CBL behaviour for norev/error cases.